clang/utils/TableGen/ClangAttrEmitter.cpp

```diff
@@ -3843,19 +3844,74 @@ void EmitClangAttrSpellingListIndex(const RecordKeeper &Records,
     const Record &R = *I.second;
     std::vector<FlattenedSpelling> Spellings = GetFlattenedSpellings(R);
     OS << "  case AT_" << I.first << ": {\n";
-    for (unsigned I = 0; I < Spellings.size(); ++ I) {
-      OS << "    if (Name == \"" << Spellings[I].name() << "\" && "
-         << "getSyntax() == AttributeCommonInfo::AS_" << Spellings[I].variety()
-         << " && Scope == \"" << Spellings[I].nameSpace() << "\")\n"
-         << "        return " << I << ";\n";
+
+    // If there are none or one spelling to check, resort to the default
+    // behavior of returning index as 0.
+    if (Spellings.size() <= 1) {
+      OS << "    return 0;\n"
+         << "    break;\n"
+         << "  }\n";
+      continue;
     }
 
-    OS << "    break;\n";
-    OS << "  }\n";
+    std::vector<StringRef> Names;
+    llvm::transform(Spellings, std::back_inserter(Names),
+                    [](const FlattenedSpelling &FS) { return FS.name(); });
+    llvm::sort(Names);
+    Names.erase(llvm::unique(Names), Names.end());
+
+    for (const auto &[Idx, FS] : enumerate(Spellings)) {
+      if (Names.size() == 1) {
+        OS << "    if (getSyntax() == AttributeCommonInfo::AS_" << FS.variety()
+           << " && ComputedScope == ";
+
+        if (FS.nameSpace() == "")
+          OS << "AttributeCommonInfo::Scope::NONE";
+        else
+          OS << "AttributeCommonInfo::Scope::" + FS.nameSpace().upper();
+
+        OS << ")\n"
+           << "      return " << Idx << ";\n";
+      } else {
+        SmallVector<StringRef, 6> SameLenNames;
+        llvm::copy_if(
+            Names, std::back_inserter(SameLenNames),
+            [&](StringRef N) { return N.size() == FS.name().size(); });
+
+        if (SameLenNames.size() == 1) {
+          OS << "    if (Name.size() == " << FS.name().size()
+             << " && getSyntax() == AttributeCommonInfo::AS_" << FS.variety()
+             << " && ComputedScope == ";
```

Comment:
Everything here and below is the same in boht branches, could you try to combine the two/have the conditional only cover the differences?  Same above.